src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,275 @@
+    // a cross-fade factor was sent to mixer.c based on time since the flight mode request was terminated
+    // it was removed primarily to simplify this PR
+
+    // hence in this PR's the user's throttle needs to be close to the hover throttle value on exiting altHold
+    // its not so bad because the 'target adjustment' by throttle requires that
+    // user throttle must be not more than half way out from hover for a stable hold
+}
+
+void altHoldUpdateTargetAltitude(void)
+{
+    // The user can raise or lower the target altitude with throttle up;  there is a big deadband.
+    // Max rate for climb/descend is 1m/s by default (up to 2.5 is allowed but overshoots a fair bit)
+    // If set to zero, the throttle has no effect.
+
+    // Some people may not like throttle being able to change the target altitude, because:
+    // - with throttle adjustment, hitting the switch won't always hold altitude if throttle is bumped
+    // - eg if the throttle is bumped low accidentally, quad will start descending.
+    // On the plus side:
+    // - the pilot has control nice control over altitude, and the deadband is wide
+    // - Slow controlled descents are possible, eg for landing
+    // - fine-tuning height is possible, eg if there is slow sensor drift
+    // - to keep the craft stable, throttle must be in the deadband, making exits smoother
+
+    const float rcThrottle = rcCommand[THROTTLE];
+
+    const float lowThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MIN); // halfway between hover and MIN, e.g. 1150 if hover is 1300
+    const float highThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MAX); // halfway between hover and MAX, e.g. 1650 if hover is 1300
+    
+    float throttleAdjustmentFactor = 0.0f;
+    if (rcThrottle < lowThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    // if failsafe is active, and we get here, we are in failsafe landing mode
+    if (failsafeIsActive()) {
+        // descend at up to 10 times faster when high
+        // default landing time is now 60s; need to get the quad down in this time from reasonable height
+        // need a rapid descent if high to avoid that timeout, and must slow down closer to ground
+        // this code doubles descent rate at 20m, to max 10x (10m/s on defaults) at 200m
+        // user should be able to descend within 60s from around 150m high without disarming, on defaults
+        // the deceleration may be a bit rocky if it starts very high up
+        throttleAdjustmentFactor = -1.0f - constrainf(altHoldState.measuredAltitudeCm * 0.0005f, 0.0f, 9.0f);
+    }
+
+    altHoldState.targetAltitudeAdjustRate = throttleAdjustmentFactor * altholdConfig()->alt_hold_target_adjust_rate * taskIntervalSeconds;
```

Comment:
Thanks, this is interesting.

I agree that the failsafe function, with it's fixed timeout, anticipates a certain descent rate.  If the user was to make it very slow, eg they were using the position hold for landing, the descent could take too long.  And vice versa, if they set a very fast descent rate, it could fall so fast with the gain we have at altitude, that we cannot arrest the descent.

However they may want to tweak the descent rate in failsafe.

If we have a 'landing mode' separately again, with its own descent rate, we will have yet another descent rate value.
I agree that it's most likely best to lock the failsafe descent rate.

Just for now, it helps me with testing if it is adjustable - I'll lock it later.

Maybe we need an `#ifdef AUTONOMOUS` label, for building purposes, instead of the user having to choose GPS Rescue, Altitude Hold, Landing Mode, Position Hold, Altitude Limit, Geofence, etc, to be included in the build.  They all share common requirements (must have GPS and/or Baro, at least).  It will get very cumbersome if each has to check whether the required code was included.

There is a similar problem with the allowed landing time, probably that has to be set constant here, since if I default it to 60s but the user has no controlled descent capability (no baro or GPS) then they have only the old code, and a long default (60s) drift time may not be appropriate.  I was thinking that in this failsafe setting that the timeout could be set to min 60s, when controlled descent is possible, but that also isn't without considerations of its own.